pkg/table-rule-selector/trie_selector.go

```diff
@@ -0,0 +1,461 @@
+		// find matched rules in table level
+		t.matchNode(si, table, matchedTableResult)
+		rules = append(rules, matchedTableResult.rules...)
+	}
+	// not found matched rule in table level, return mathed rule in schema level
+	t.addToCache(cacheKey, rules)
+	return rules.clone()
+}
+
+// Remove implements Selector interface.
+// TODO: remove useless nodes and lazy deletion
+func (t *trieSelector) Remove(schema, table string) error {
+	t.Lock()
+	defer t.Unlock()
+
+	if len(schema) == 0 {
+		return errors.NotValidf("schema pattern %s", schema)
+	}
+
+	schemaItems, err := t.track(t.root, schema)
+	if err != nil {
+		return errors.Annotate(err, "schema level")
+	}
+
+	schemaLeafItem := schemaItems[len(schemaItems)-1]
+	if len(table) > 0 {
+		if schemaLeafItem.nextLevel == nil {
+			return errors.NotFoundf("table level %s", table)
+		}
+
+		tableItems, err := t.track(schemaLeafItem.nextLevel, table)
+		if err != nil {
+			return errors.Annotate(err, "table level")
+		}
+
+		if tableItems[len(tableItems)-1].rule == nil {
+			return errors.NotFoundf("pattern schema %s table %s", schema, table)
+		}
+
+		// remove table level nodes
+		tableItems[len(tableItems)-1].rule = nil
+		t.clearCache()
+		return nil
+	}
+
+	if schemaLeafItem.rule == nil {
+		return errors.NotFoundf("pattern schema %s", schema)
+	}
+
+	schemaLeafItem.rule = nil
+	t.clearCache()
+	return nil
+}
+
+func (t *trieSelector) track(n *node, pattern string) ([]*item, error) {
+	items := make([]*item, 0, len(pattern))
+	for i := range pattern {
+		switch pattern[i] {
+		case asterisk:
+			if n.asterisk == nil {
+				return nil, errors.NotFoundf("pattern %v", pattern)
+			}
+
+			if i != len(pattern)-1 {
+				return nil, errors.NotValidf("pattern %v ", pattern)
+			}
+
+			items = append(items, n.asterisk)
+		case question:
+			if n.question == nil {
+				return nil, errors.NotFoundf("pattern %v", pattern)
+			}
+
+			items = append(items, n.question)
+			n = n.question.child
+		default:
+			item, ok := n.characters[pattern[i]]
+			if !ok {
+				return nil, errors.NotFoundf("pattern %v", pattern)
+			}
+			items = append(items, item)
+			n = item.child
+		}
+	}
+
+	return items, nil
+}
+
+// AllRules implements Router's AllRules
+func (t *trieSelector) AllRules() (map[string]interface{}, map[string]map[string]interface{}) {
+	var (
+		tableRules  = make(map[string]map[string]interface{})
+		schemaNodes = make(map[string]*node)
+		schemaRules = make(map[string]interface{})
+		characters  []byte
+	)
+	t.RLock()
+	t.travel(t.root, characters, schemaRules, schemaNodes)
+
+	for schema, n := range schemaNodes {
+		rules, ok := tableRules[schema]
+		if !ok {
+			rules = make(map[string]interface{})
+		}
+
+		characters = characters[:0]
+		t.travel(n, characters, rules, nil)
+		if len(rules) > 0 {
+			tableRules[schema] = rules
+		}
+	}
+	t.RUnlock()
+	return schemaRules, tableRules
+}
+
+func (t *trieSelector) travel(n *node, characters []byte, rules map[string]interface{}, nodes map[string]*node) {
```

Comment:
confuse `characters`  with `node.characters`, how about rename it? like `prefixChars`

Reply:
It's not prefixChars, rename it to word